feat: Convert predicate to arrow filter and push down to parquet reader

viirya opened the pull request:
close #265

viirya commented on crates/iceberg/src/arrow.rs:
```diff
@@ -113,6 +143,405 @@ impl ArrowReader {
+                    is_not_null(column)
+                },
+            ))),
+            PredicateOperator::IsNan => {
+                todo!("IsNan is not supported yet")
+            }
+            PredicateOperator::NotNan => {
+                todo!("NotNan is not supported yet")
+            }
```
Seems there are no existing kernels for is_nan and not_nan operators. I've not added them in this PR and will implement them in following work.

viirya commented on crates/iceberg/src/arrow.rs:
```diff
@@ -113,6 +143,405 @@ impl ArrowReader {
+
+    fn visit_set(&mut self, predicate: &SetExpression<BoundReference>) -> Result<Self::T> {
+        match predicate.op() {
+            PredicateOperator::In => {
+                todo!("In is not supported yet")
+            }
+            PredicateOperator::NotIn => {
+                todo!("NotIn is not supported yet")
+            }
```
ditto.

liurenjie1024 commented:
cc @viirya Is this ready for review or you still need to do more update?

viirya commented:
@liurenjie1024 It is ready for review. I will fix the conflicts.

liurenjie1024 commented on crates/iceberg/src/arrow.rs:
```diff
@@ -20,24 +20,38 @@
     batch_size: Option<usize>,
     file_io: FileIO,
     schema: SchemaRef,
+    predicates: Option<Vec<BoundPredicate>>,
```
The `Vec` is kind of confusing to me, is it conjunction or disjunction? Since we already have `And/Or` as part of `BoundPredicate` variant, how about just `BoundPredicate`？

viirya replied:
This is because the Parquet API `RowFilter` takes `predicates: Vec<Box<dyn ArrowPredicate>>`.

This is `RowFilter`'s doc:

```
/// A [`RowFilter`] allows pushing down a filter predicate to skip IO and decode
///
/// This consists of a list of [`ArrowPredicate`] where only the rows that satisfy all
/// of the predicates will be returned. 
```

So I think it is conjunction relationships between these predicates.

> Since we already have And/Or as part of BoundPredicate variant, how about just BoundPredicate？

Yea, I can use just one `BoundPredicate`. So users can define conjunctions in one single predicate using `And`.

liurenjie1024 commented on crates/iceberg/src/arrow.rs:
```diff
@@ -113,6 +143,405 @@ impl ArrowReader {
+}
+
+/// A visitor for bound predicates.
+pub trait BoundPredicateVisitor {
```
How about moving this to `expr/predicate`  module?

viirya replied:
Okay

liurenjie1024 commented on crates/iceberg/src/arrow.rs:
```diff
@@ -113,6 +143,405 @@ impl ArrowReader {
+    type U;
+
+    /// Visit a bound predicate.
+    fn visit_predicate(&mut self, predicate: &BoundPredicate) -> Result<Self::T> {
```
Though it's correct, it's not typical post order visitor pattern, how about moving this travels logic into a standalone   funtion, so that we can follow convention of other visitors?

liurenjie1024 commented on crates/iceberg/src/arrow.rs:
```diff
@@ -113,6 +143,405 @@ impl ArrowReader {
+        PrimitiveLiteral::Long(value) => Box::new(Int64Array::new_scalar(*value)),
+        PrimitiveLiteral::Float(value) => Box::new(Float32Array::new_scalar(value.as_f32())),
+        PrimitiveLiteral::Double(value) => Box::new(Float64Array::new_scalar(value.as_f64())),
+        _ => todo!("Unsupported literal type"),
```
We should not panic here, but throws error.

viirya replied:
Currently I cannot return an error from `get_arrow_datum` to propagate an error. The closure needs to get the ownership of all referenced variables. But we cannot call `get_arrow_datum` outside the closure to propagate the error and pass `Datum` into the closure because `Datum` is not implemented with `Send`. I proposed https://github.com/apache/arrow-rs/pull/5587 and see if we can get it.

liurenjie1024 commented on crates/iceberg/src/arrow.rs:
```diff
@@ -113,6 +143,405 @@ impl ArrowReader {
+}
+
+/// A visitor to collect field ids from bound predicates.
+struct CollectFieldIdVisitor {
```
Could we add some ut for this?

viirya replied:
OKay

liurenjie1024 commented on crates/iceberg/src/arrow.rs:
```diff
@@ -113,6 +143,405 @@ impl ArrowReader {
+                },
+            ))),
+            PredicateOperator::IsNan => {
+                todo!("IsNan is not supported yet")
```
We should not panic here, it should return true.

liurenjie1024 commented on crates/iceberg/src/arrow.rs:
```diff
@@ -113,6 +143,405 @@ impl ArrowReader {
+            }
+            op => Err(Error::new(
+                ErrorKind::DataInvalid,
+                format!("Unsupported unary operator: {op}"),
```
We should not return error, but always true.

viirya replied:
Ah, good point.

liurenjie1024 commented on crates/iceberg/src/arrow.rs:
```diff
@@ -113,6 +143,405 @@ impl ArrowReader {
+            PredicateOperator::NotNull => Ok(Box::new(ArrowPredicateFn::new(
+                self.projection_mask.clone(),
+                move |batch| {
+                    let column = batch.column(term_index);
```
This maybe incorrect for nested column, I think maybe we should either return projection_mask for each leave column, or implement a general purpose flatten method for struct array.

viirya replied:
The Parquet reader API will flatten required columns based on the projection_mask we provide. I.e., If the projection mask selects one nested column `a.b`, it will be the first column of the record batch when calling `evaluate` of `ArrowPredicate` API.

liurenjie1024 replied:
I did some test but it seems that it doesn't work like this, say the schem is like following:
```
message {
  struct a {
    int b
  }
}
```
And we pass `ProjectionMask::leaves([0])`, it will return struct array, so `batch.column[term_index]` will return `StructArray`

viirya replied:
Hmm, that's what I got from reading the Parquet code. Let me take further look and test.

viirya replied:
Just wrote a test to test it in arrow-rs. Yea, for a nested column such as `struct.a`, the batch passed to `evaluate` contains a struct array with the column `a`. Looks like Parquet will project the requested column indices and construct the upper nested type (i.e., `struct`) before passing to `evaluate`,

viirya replied:
Added the test to arrow-rs to clarify its usage: https://github.com/apache/arrow-rs/pull/5600

viirya replied:
I think to have a flatten method for struct array sounds more simple way. I'm looking in arrow-rs to see if there is existing one, if not, we need to implement it here.

viirya replied:
> This maybe incorrect for nested column, I think maybe we should either return projection_mask for each leave column, or implement a general purpose flatten method for struct array.

I tried to change to return projection_mask for each leave column, it is pretty straightforward to implement. Please let me know if it looks good to you. Thanks.

liurenjie1024 reviewed (commented):
Hi, @viirya Thanks for this pr, the general idea of reusing arrow kernel looks great! But  I found some small problems which could be improved.

viirya commented:
I've addressed some of above reviews. I will resolve other reviews soon. Thanks.

viirya commented on crates/iceberg/src/arrow/reader.rs:
```diff
@@ -186,4 +216,399 @@ impl ArrowReader {
+            let struct_array = column.as_any().downcast_ref::<StructArray>().unwrap();
+            get_leaf_column(struct_array.column(0))
+        }
+        _ => Ok(column.clone()),
```
I'm not sure if `BoundReference` could be referred to nested field inside list or map in Iceberg. Assume that it is either struct or primitive now.

liurenjie1024 replied:
I also think it doesn't support it now.

liurenjie1024 commented on crates/iceberg/src/arrow/reader.rs:
```diff
@@ -186,4 +216,399 @@ impl ArrowReader {
+    field_ids: Vec<i32>,
+}
+
+impl BoundPredicateVisitor for CollectFieldIdVisitor {
```
#334 has got merged, do you mind to rewrite this?

viirya replied:
Okay, let me take a look, not sure if it is completely suitable for the usage here, especially for `PredicateConverter`. But let me try it.

viirya replied:
Rewrote it using the API.

liurenjie1024 commented on crates/iceberg/src/arrow/reader.rs:
```diff
@@ -186,4 +216,399 @@ impl ArrowReader {
+
+            // Collect Parquet column indices from field ids
+            let mut collector = CollectFieldIdVisitor { field_ids: vec![] };
+            visit_predicate(&mut collector, predicates).unwrap();
```
We don't need to do this for every file?

viirya replied:
Hmm, but we can only get `SchemaDescriptor` from the parquet reader for the each file.

viirya replied:
Oh, I got your point. I moved `CollectFieldIdVisitor` out of the loop of `tasks`.

liurenjie1024 commented on crates/iceberg/src/arrow/reader.rs:
```diff
@@ -186,4 +216,399 @@ impl ArrowReader {
+                .field_ids
+                .iter()
+                .map(|field_id| {
+                    field_id_map.get(field_id).cloned().ok_or_else(|| {
```
This doesn't have to be an error? Missing some field maybe valid due to schema evolution.

viirya replied:
Hmm, if the predicate uses a field not in schema, even with schema evaluation, shouldn't it be considered an error?

E.g., given a predicate `a == 1`, but there is no `a` column.

liurenjie1024 replied:
I think this is possible. For example, in version 1 iceberg table has no column a, and we inserted some data. Then we upgraded the schema to add column a, and it becomes version 2. Since iceberg is lazy, parquets files in version 1 has no column a.

viirya replied:
Okay. Removed the error and just ignored it.

liurenjie1024 commented on crates/iceberg/src/arrow/reader.rs:
```diff
@@ -186,4 +216,399 @@ impl ArrowReader {
+    }
+
+    /// Build the map of field id to Parquet column index in the schema.
+    fn build_field_id_map(&self, parquet_schema: &SchemaDescriptor) -> Result<HashMap<i32, usize>> {
```
```suggestion
    fn build_field_id_map(parquet_schema: &SchemaDescriptor) -> Result<HashMap<i32, usize>> {
```
I think we don't need `self` parameter?

viirya replied:
Yea, removed it.

liurenjie1024 commented on crates/iceberg/src/arrow/reader.rs:
```diff
@@ -186,4 +216,399 @@ impl ArrowReader {
+    }
+}
+
+impl<'a> BoundPredicateVisitor for PredicateConverter<'a> {
```
Do you mind to rewrite it using new api?

viirya replied:
Rewrote it.

liurenjie1024 reviewed (commented):
Thanks @viirya for this pr, it looks promising! Given #334 has been merged, do you mind to rewrite some of the visitors with new api? This helps to make code easier to maintain.

viirya commented:
@liurenjie1024 Thanks for review. Sorry for late. I addressed the comments by rewriting the visitors using the new API. I replied with another questions.

liurenjie1024 commented on crates/iceberg/src/arrow/reader.rs:
```diff
@@ -186,4 +221,637 @@ impl ArrowReader {
+                    return Err(Error::new(
+                        ErrorKind::DataInvalid,
+                        format!(
+                            "Leave column {:?} in schema doesn't have field id",
```
How about adding `basic_info.name` and index into error message to make it better to debug?

viirya replied:
Okay

liurenjie1024 commented on crates/iceberg/src/arrow/reader.rs:
```diff
@@ -186,4 +221,637 @@ impl ArrowReader {
+impl BoundPredicateVisitor for CollectFieldIdVisitor {
+    type T = ();
+
+    fn always_true(&mut self) -> Result<Self::T> {
```
As we have discussed before, it's better to use concrete types rather than `Self::T` to make it easier to read.

liurenjie1024 commented on crates/iceberg/src/arrow/reader.rs:
```diff
@@ -186,4 +221,637 @@ impl ArrowReader {
+    fn always_true(&mut self) -> Result<Self::T> {
+        Ok(Box::new(ArrowPredicateFn::new(
+            self.projection_mask.clone(),
+            |batch| Ok(BooleanArray::from(vec![true; batch.num_rows()])),
```
This maybe not a blocker, but is it possible to build a const array in arrow?

viirya replied:
As I know, there is no const array in arrow for now. It has dictionary array which I think is close to const array, but the `ArrowPredicateFn` API defines returned type to be `BooleanArray`.

liurenjie1024 commented on crates/iceberg/src/arrow/reader.rs:
```diff
@@ -186,4 +221,637 @@ impl ArrowReader {
+        // Find the column index in projection mask.
+        let column_idx = self
+            .columns
+            .iter()
```
Why not just `Vec::get` ? This iteration needs to go through all columns.

viirya replied:
Here we want to get the vector index of the Parquet leaf column index. For example, assume the vector is vector![1, 0, 3, 2], and the Parquet column index is 3, we need to get 2.

viirya replied:
Actually this looking up is not necessary. Removed it.

liurenjie1024 commented on crates/iceberg/src/arrow/reader.rs:
```diff
@@ -186,4 +221,637 @@ impl ArrowReader {
+        reference: &BoundReference,
+        _predicate: &BoundPredicate,
+    ) -> Result<Self::T> {
+        let projected_mask = self.bound_reference(reference)?;
```
This seems incorrect to me. Let's say the predicate is `a is null AND b >1`, then the batch passed to this `ArrowPredicateFn` is constructed by projection mask of [a, b]. I think one possible solution is to use same project mask for all predicates, and pass the column_idx to get_leaf_column.

viirya replied:
Hmm, as `RowFilter` API design accepts multiple `ArrowPredicate`s. Each `ArrowPredicate` has its `projection` and the API doc said the API will be passed batches that contains the columns specified in `projection`. I think it should be projected for each `ArrowPredicate` and `projection`.

viirya replied:
I will go to test it in arrow-rs to verify it.

viirya replied:
Ah, no, although I used multiple `ArrowPredicates` before, it was changed to one `ArrowPredicates` after you suggested. So now we generate only one `ArrowPredicates`.

Hmm, let me think how do deal with it.

viirya replied:
Based on what I searched and the kindly reply on the issue, I think there is no way to do nested projection on RecordBatch currently.

To implement the feature in arrow-rs might block this. I tend to finish top-level column only in this PR.

WDYT, @liurenjie1024 ?

liurenjie1024 commented on crates/iceberg/src/arrow/reader.rs:
```diff
@@ -186,4 +221,637 @@ impl ArrowReader {
+        PrimitiveLiteral::Float(value) => Ok(Box::new(Float32Array::new_scalar(value.as_f32()))),
+        PrimitiveLiteral::Double(value) => Ok(Box::new(Float64Array::new_scalar(value.as_f64()))),
+        l => Err(Error::new(
+            ErrorKind::DataInvalid,
+            format!("Unsupported literal type: {:?}", l),
```
```suggestion
            ErrorKind::FeatureUnsupported,
            format!("Converting datum from type {:?} to arrow not supported yet.", l),
```

liurenjie1024 commented on crates/iceberg/src/arrow/reader.rs:
```diff
@@ -186,4 +221,637 @@ impl ArrowReader {
+    }
+}
+
+fn get_arrow_datum(datum: &Datum) -> Result<Box<dyn ArrowDatum + Send>> {
```
Do you mind to move this to arrow module? I guess this would be quite useful and we could add more type support late.

viirya replied:
This is already in arrow module. Do you mean schema.rs?

viirya replied:
Moved to `schema.rs`.

liurenjie1024 commented on crates/iceberg/src/arrow/reader.rs:
```diff
@@ -186,4 +221,637 @@ impl ArrowReader {
+impl<'a> BoundPredicateVisitor for PredicateConverter<'a> {
+    type T = Box<dyn ArrowPredicate>;
+
+    fn always_true(&mut self) -> Result<Self::T> {
```
Please use concrete type here to improve readability😅

viirya replied:
Sure. I probably missed previous comment.

liurenjie1024 commented on crates/iceberg/src/arrow/reader.rs:
```diff
@@ -186,4 +221,637 @@ impl ArrowReader {
+
+/// A visitor to collect field ids from bound predicates.
+struct CollectFieldIdVisitor {
+    field_ids: Vec<i32>,
```
Should this be a set?

viirya replied:
Yea, changed to a set.

liurenjie1024 commented on crates/iceberg/src/scan.rs:
```diff
@@ -689,4 +720,90 @@ mod tests {
+    }
+
+    #[tokio::test]
+    async fn test_filter_on_arrow_is_not_null() {
```
Thanks for the tests, is it possible to add serveral test cases for more complex types such as `AND`, `OR`?

viirya replied:
Yes. I added some more tests using `AND` and `OR`.

liurenjie1024 reviewed (commented):
Thanks @viirya for this great pr! It's really hard work, and you did it in a quite elegant way. I found some small problems to fix.

viirya commented on crates/iceberg/src/arrow/reader.rs:
```diff
@@ -186,4 +219,634 @@ impl ArrowReader {
+                },
+            )))
+        } else {
+            self.build_always_true()
```
Yes, fixed it.

viirya commented on crates/iceberg/src/arrow/reader.rs:
```diff
@@ -186,4 +219,634 @@ impl ArrowReader {
+                },
+            )))
+        } else {
+            self.build_always_true()
```
Okay

viirya commented on crates/iceberg/src/arrow/reader.rs:
```diff
@@ -186,4 +219,634 @@ impl ArrowReader {
+                Ok(BooleanArray::from(vec![true; batch.num_rows()]))
+            })))
+        } else {
+            self.build_always_true()
```
Okay

viirya commented:
@liurenjie1024 I've addressed your comments. Please take a look when you can. Thanks.

viirya commented:
Thanks @liurenjie1024. I just resolved the conflicts.

viirya commented:
Thanks @liurenjie1024 for your review!
